File: html.php
<?php

function text(string $str): string {
	return str_replace(['<', '>', '"', '&'], ['&lt;', '&gt;', '&quot;', '&amp;'], $str);
}

function node(string $tag, string ...$nodes): string {
	$t = implode('', $nodes);
	return "<$tag>$t</$tag>";
}

function tnode(string $tag, string $t): string {
	$t = text($t);
	return "<$tag>$t</$tag>";
}

function attributes(array $attr): string {
	$a = '';
	foreach ($attr as $key => $val) {
		$a .= ' ' . $key;
		if ($val != '') {
			$a .= "=\"" . text($val) . "\"";
		}
	}
	return $a;
}

function anode(string $tag, array $attr = [], string ...$nodes): string {
	$a = attributes($attr);
	$t = implode('', $nodes);
	return "<$tag$a>$t</$tag>";
}

function inode(string $tag, array $attr = []): string {
	$a = attributes($attr);
	return "<$tag$a>";
}

function nodes(string ...$nodes): string {
	return implode('', $nodes);
}


function milliseconds() {
	$mt = explode(' ', microtime());
	return intval( $mt[1] * 1E3 ) + intval( round( $mt[0] * 1E3 ) );
}
$start = milliseconds();
$html = '';
for ($i=0; $i<100_000; $i++) {
	$html = nodes(
		inode('!doctype', ['html'=>'']),
		anode('html', ['lang'=>'en'], nodes(
			node('head',
				inode('meta', ['charset'=>'utf-8']),
				tnode('title', 'Test'),
			),
			node('body',
				anode('div', ['class'=>'alert alert-danger'], text('Test')),
				tnode("span", 'Abc & def'),
			),
		)),
	);
}
$end = milliseconds();
$ms = $end-$start;
echo $html."\n";
echo "PHP: ${ms}ms\n";

File: html.nim
import tables
import strutils
import times

# type attr = table[string, string]

proc text(str: string): string =
    result = str.multiReplace(("&", "&amp;"), ("<", "&lt;"), (">", "&gt;"), ("\"", "&quot;"))


proc node(tag: string, nodes: varargs[string]): string =
    result = "<" & tag & ">" & join(nodes, "") & "</" & tag & ">"

proc tnode(tag: string, t: string): string =
    result = "<" & tag & ">" & text(t) & "</" & tag & ">"

proc attributes(attr: Table[string, string]): string =
    result = ""
    for key, val in attr:
        result = result & " " & key
        if val != "":
            result = result & "=\"" & text(val) & "\""

proc anode(tag: string, attr: Table[string, string], nodes: varargs[string]): string =
    result = "<" & tag & attributes(attr) & ">"
    for node in nodes:
        result = result & node
    result = result & "</" & tag & ">"

proc inode(tag: string, attr: Table[string, string]): string =
    result = "<" & tag & attributes(attr) & ">"

proc nodes(nodes: varargs[string]): string =
    result = ""
    for node in nodes:
        result = result & node

var tstart = cpuTime()
var html = ""
for i in countup(0, 100_000):
    html = nodes(
        inode("!doctype", {"html": ""}.toTable),
        anode("html", {"lang": "en"}.toTable,
            node("head",
                inode("meta", {"charset": "utf-8"}.toTable),
                tnode("title", "Test"),
            ),
            node("body",
                anode("div", {"class": "alert alert-danger"}.toTable, text("Test")),
                tnode("span", "Abc & def"),
            ),
        ),
    )
var tend = cpuTime()
echo html, "\n"
echo "Nim: ", tend - tstart, "\n"
